File: backend/app/models/product.py
from datetime import datetime
from sqlalchemy import String, DateTime, Float, Integer
from sqlalchemy.orm import Mapped, mapped_column
from app.core.database import Base
from sqlalchemy import Computed


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    n_pct: Mapped[float] = mapped_column(Float, default=0.0)
    p_pct: Mapped[float] = mapped_column(Float, default=0.0)
    k_pct: Mapped[float] = mapped_column(Float, default=0.0)
    ca_pct: Mapped[float] = mapped_column(Float, default=0.0)
    mg_pct: Mapped[float] = mapped_column(Float, default=0.0)
    s_pct: Mapped[float] = mapped_column(Float, default=0.0)
    fe_pct: Mapped[float] = mapped_column(Float, default=0.0)
    cu_pct: Mapped[float] = mapped_column(Float, default=0.0)
    mn_pct: Mapped[float] = mapped_column(Float, default=0.0)
    b_pct: Mapped[float] = mapped_column(Float, default=0.0)
    zn_pct: Mapped[float] = mapped_column(Float, default=0.0)
    weight_lbs: Mapped[float] = mapped_column(Float, nullable=True)
    cost_per_bag: Mapped[float | None] = mapped_column(Float, nullable=True)
    sgn: Mapped[str] = mapped_column(String(32), nullable=True)
    product_link: Mapped[str] = mapped_column(String(512), nullable=True)
    label: Mapped[str] = mapped_column(String(512), nullable=True)
    sources: Mapped[str] = mapped_column(String, nullable=True)
    urea_nitrogen: Mapped[float] = mapped_column(Float, nullable=True)
    ammoniacal_nitrogen: Mapped[float] = mapped_column(Float, nullable=True)
    water_insol_nitrogen: Mapped[float] = mapped_column(Float, nullable=True)
    other_water_soluble: Mapped[float] = mapped_column(Float, nullable=True)
    slowly_available_from: Mapped[str] = mapped_column(String(255), nullable=True)
    cost_per_lb_n: Mapped[float] = mapped_column(
        Float,
        Computed(
            "CASE WHEN n_pct IS NOT NULL AND n_pct != 0 AND weight_lbs IS NOT NULL AND weight_lbs != 0 "
            "THEN cost_per_bag / ((n_pct / 100) * weight_lbs) ELSE NULL END",
            persisted=True,
        ),
        nullable=True,
    )
    cost_per_lb: Mapped[float] = mapped_column(
        Float,
        Computed(
            "CASE WHEN weight_lbs IS NOT NULL AND weight_lbs != 0 "
            "THEN cost_per_bag / weight_lbs ELSE NULL END",
            persisted=True,
        ),
        nullable=True,
    )
    last_scraped_price: Mapped[float] = mapped_column(Float, nullable=True)
    last_scraped_at: Mapped[datetime] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )

File: frontend/src/types/product.ts
export interface Product {
  id: number;
  name: string;
  description?: string;
  n_percent: number;
  p_percent: number;
  k_percent: number;
  ca_percent?: number;
  mg_percent?: number;
  s_percent?: number;
  fe_percent?: number;
  cu_percent?: number;
  mn_percent?: number;
  b_percent?: number;
  zn_percent?: number;
  cost_per_lb: number;
  cost_per_lb_n: number;
  created_at: string;
  updated_at: string;
}

// This type should match all fields used in the ProductForm component
export type ProductFormValues = {
  name: string;
  product_link?: string;
  weight_lbs?: number;
  cost_per_bag?: number;
  sgn?: string;
  label?: string;
  sources?: string;
  n_pct?: number;
  p_pct?: number;
  k_pct?: number;
  ca_pct?: number;
  mg_pct?: number;
  s_pct?: number;
  fe_pct?: number;
  cu_pct?: number;
  mn_pct?: number;
  b_pct?: number;
  zn_pct?: number;
  urea_nitrogen?: number;
  ammoniacal_nitrogen?: number;
  water_insol_nitrogen?: number;
  other_water_soluble?: number;
  slowly_available_from?: string;
  last_scraped_price?: number;
  last_scraped_at?: string;
};
